presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,174 @@
+                .collect(toSet());
+
+        Assignments.Builder childAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : child.getAssignments().entrySet()) {
+            if (!targets.contains(assignment.getKey())) {
+                childAssignments.put(assignment);
+            }
+        }
+        for (Symbol input : inputs) {
+            childAssignments.putIdentity(input);
+        }
+
+        return Optional.of(
+                new ProjectNode(
+                        parent.getId(),
+                        new ProjectNode(
+                                child.getId(),
+                                child.getSource(),
+                                childAssignments.build()),
+                        Assignments.copyOf(parentAssignments)));
+    }
+
+    private Expression inlineReferences(Expression expression, Assignments assignments)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Object>()
+        {
+            @Override
+            public Expression rewriteSymbolReference(SymbolReference reference, Object context, ExpressionTreeRewriter treeRewriter)
+            {
+                Symbol symbol = Symbol.from(reference);
+                Expression result = assignments.get(symbol);
+                if (result != null) {
+                    return result;
+                }
+
+                return reference;
+            }
+        }, expression);
+    }
+
+    private Sets.SetView<Symbol> extractInliningTargets(ProjectNode parent, ProjectNode child)
+    {
+        // candidates for inlining are
+        //   1. references to simple constants
+        //   2. references to complex expressions that
+        //      a. are not inputs to try() expressions
+        //      b. appear only once across all expressions
+        //      c. are not identity projections
+
+        Map<Symbol, Long> dependencies = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+        // find references to simple constants
+        Set<Symbol> constants = dependencies.keySet().stream()
+                .filter(input -> child.getAssignments().get(input) instanceof Literal)
+                .collect(toSet());
+
+        // exclude any complex inputs to TRY expressions. Inlining them would potentially
+        // change the semantics of those expressions
+        Set<Symbol> tryArguments = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> extractTryArguments(e).stream())
+                .collect(toSet());
+
+        Set<Symbol> singletons = dependencies.entrySet().stream()
+                .filter(e -> e.getValue() == 1) // reference appears just once across all expressions in parent project node
```

Comment:
Except for simple iteration variables in `for` and `while` loops. Lambdas are still one of those places I'm still on the fence about. There are cases where a simple variable name looks more readable (`e` for element or entry, especially when the operation done by the lambda is straightforward). 

I'll make the change here, though, and see how it looks.